mapper/pojo-standalone/src/main/java/org/hibernate/search/mapper/pojo/standalone/loading/impl/StandalonePojoMassIdentifierLoader.java

```diff
@@ -5,13 +5,27 @@
 package org.hibernate.search.mapper.pojo.standalone.loading.impl;
 
 import org.hibernate.search.mapper.pojo.loading.spi.PojoMassIdentifierLoader;
+import org.hibernate.search.mapper.pojo.loading.spi.PojoMassIdentifierLoadingContext;
 import org.hibernate.search.mapper.pojo.standalone.loading.MassIdentifierLoader;
+import org.hibernate.search.mapper.pojo.standalone.loading.MassIdentifierSimpleLoader;
 
-public class StandalonePojoMassIdentifierLoader implements PojoMassIdentifierLoader {
+abstract class StandalonePojoMassIdentifierLoader<T extends MassIdentifierSimpleLoader> implements PojoMassIdentifierLoader {
 
-	private final MassIdentifierLoader delegate;
+	protected final T delegate;
 
-	public StandalonePojoMassIdentifierLoader(MassIdentifierLoader delegate) {
+	public static <T extends MassIdentifierSimpleLoader> StandalonePojoMassIdentifierLoader<?> loader(T delegate,
+			StandalonePojoMassIdentifierSink<?> sink,
+			PojoMassIdentifierLoadingContext<?> context) {
+		if ( delegate instanceof MassIdentifierLoader ) {
```

Comment:
I would personally avoid features based on casts and `instanceof` checks, because they are easily missed by implementers, and requires them to actively look for the feature in order to find it.

I and would personally favor composition and/or explicit method signature.

For example, you could:

1. Add a new method to `MassIndexingStrategy` to *optionally* provide something that can count entities to index: `Optional<MassEntityCounter<...>> createCounter(...)`.
2. OR add change the method in `MassIdentifierLoader` to *optionally* return the total count: `Optional<Long> totalCount()`.
3. OR a mix of the above: add a new method to `MassIdentifierLoader` to *optionally* provide something that can count entities to index: `Optional<MassEntityCounter<...>> counter()`.
4. ...

Let me know what you think.

Reply:
> I didn't have time to go through it all yet, but I think this comment is already enough to start a discussion, so I'll send it anyway :)

👍🏻 that's actually good, as I wasn't sure about the approach, and there was no point in looking at other things if the approach is bad 😄